app/models/miq_server/worker_management/monitor/kubernetes.rb

```diff
@@ -0,0 +1,77 @@
+module MiqServer::WorkerManagement::Monitor::Kubernetes
+  extend ActiveSupport::Concern
+    attr_accessor :pod_resource_version
+
+    def current_pods
+      # TODO: Add mutex around access
+      @current_pods ||= begin
+        # The first call to current_pods will return an empty hash since the newly created collector thread hasn't collected anything yet.
+        collector_thread
+        {}
+      end
+    end
+
+    def cleanup_failed_deployments
+      # TODO: We should have a list of worker deployments we'll delete to avoid accidentally killing pg/memcached/orchestrator
+      # See ContainerOrchestrator#get_pods
+      failed_deployments.each do |failed|
+        orchestrator.delete_deployment(failed)
+      end
+    end
+
+    def failed_deployments(restart_count = 5)
+      # TODO: This logic might flag deployments that are hitting memory/cpu limits or otherwise not really 'failed'
+      current_pods.select { |name, h| h.fetch(:last_state_terminated) && h.fetch(:container_restarts, 0) > restart_count }.collect { |name, h| h[:label_name] }
+    end
+
+    private
+    def collector_thread
+      # TODO: Ensure thread exceptions kill the thread and a new thread is created
+      @collector_thread ||= Thread.new { pod_collector }
+    end
+
+    def orchestrator
+      @orchestrator ||= ContainerOrchestrator.new
+    end
+
+    def pod_collector
+      # TODO: To ensure we're in sync, we might want to break out of the watch, reset the current_pods and run this again
+      collect_initial_pods
+      watch_for_pod_events
+    end
+
+    def collect_initial_pods
+      pods = orchestrator.get_pods
+      pods.each { |p| save_pod(p) }
+      self.pod_resource_version = pods.resourceVersion || 0
+    end
+
+    def watch_for_pod_events
+      orchestrator.watch_pods(self.pod_resource_version) do |event|
+        case event.type.downcase
+        when "added", "modified"
+          save_pod(event.object)
+        when "deleted"
+          delete_pod(event.object)
+        when "error"
+          # TODO
+        end
```

Comment:
One additional case that can happen is 410 Gone if you ask for a resourceVersion that is already purged from etcd, In the k8s RefreshWorker we check for `notice.kind == "Status" && notice.code == 410`.  Just have to do the full `get_pods` again to get the new resourceVersion.

Highely unlikely given we're in a loop and immediately passing in the resourceVersion back in but still possible

Reply:
Yeah, thanks!  I knew there were other cases I need to handle and honestly, it feels like we should just restart the thread periodically to make sure it's sync in the event of unhandled errors/situations.